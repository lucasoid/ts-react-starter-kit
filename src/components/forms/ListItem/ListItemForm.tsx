import * as React from 'react';
import uuidv4 from 'uuid/v4';
import { Form } from '~components/forms/Form';
import { TextField } from '~components/forms/Inputs';

interface IListFormProps {
    onSubmit: Function;
    onCancel: Function;
}
export const ListItemForm: React.FC<IListFormProps> = props => {
    const [name, setName] = React.useState('');
    const [category, setCategory] = React.useState('');
    const [quantity, setQuantity] = React.useState(1);

    return (
        <Form
            onSubmit={() => {
                props.onSubmit({
                    id: uuidv4(),
                    name: name,
                    category: category,
                    quantity: quantity,
                    isActive: true,
                });
            }}
            onCancel={props.onCancel}
            submitText={'Create'}
            cancelText={'Cancel'}
        >
            <TextField value={name} label="Name" onChange={setName} />
            <TextField value={category} label="Category" onChange={setCategory} />
            <TextField value={quantity} label="Quantity" onChange={setQuantity} />
        </Form>
    );
};
