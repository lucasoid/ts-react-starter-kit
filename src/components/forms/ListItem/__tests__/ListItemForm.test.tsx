import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { ListItemForm } from '../ListItemForm';

jest.mock('react', () => {
    const original = jest.requireActual('react');
    return {
        __esModule: true,
        ...original,
        useState: jest.fn(),
    };
});

describe('ListItemForm', () => {
    it('renders a form for creating a list', () => {
        const onSubmit = jest.fn();
        const onCancel = jest.fn();
        const setter = jest.fn();
        // @ts-ignore
        React.useState.mockImplementation(initialState => {
            if (typeof initialState === 'string') {
                return ['value', setter];
            }
            if (typeof initialState === 'number') {
                return [10, setter];
            }
        });
        let wrapper = shallow(<ListItemForm onSubmit={onSubmit} onCancel={onCancel} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('invokes onSubmit with some default values populated upon creation', () => {
        const onSubmit = jest.fn();
        const onCancel = jest.fn();
        const setter = jest.fn();
        // @ts-ignore
        React.useState.mockImplementation(initialState => {
            if (typeof initialState === 'string') {
                return ['value', setter];
            }
            if (typeof initialState === 'number') {
                return [10, setter];
            }
        });
        let wrapper = mount(<ListItemForm onSubmit={onSubmit} onCancel={onCancel} />);
        wrapper.find('form').simulate('submit', { e: { preventDefault: () => {} } });
        expect(onSubmit.mock.calls[0][0].id).toBeTruthy();
        expect(onSubmit.mock.calls[0][0].name).toBe('value');
        expect(onSubmit.mock.calls[0][0].category).toBe('value');
        expect(onSubmit.mock.calls[0][0].quantity).toBe(10);
        expect(onSubmit.mock.calls[0][0].isActive).toBe(true);
    });
});
