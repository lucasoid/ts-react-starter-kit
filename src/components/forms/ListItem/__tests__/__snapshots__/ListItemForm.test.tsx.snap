// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemForm renders a form for creating a list 1`] = `
<Form
  cancelText="Cancel"
  onCancel={[MockFunction]}
  onSubmit={[Function]}
  submitText="Create"
>
  <TextField
    label="Name"
    onChange={[MockFunction]}
    value="value"
  />
  <TextField
    label="Category"
    onChange={[MockFunction]}
    value="value"
  />
  <TextField
    label="Quantity"
    onChange={[MockFunction]}
    value={10}
  />
</Form>
`;
